enhancements/kubelet/oci-image-volume.md

```diff
@@ -0,0 +1,180 @@
+- Mounting thousands of images in a single pod.
+
+## Proposal
+
+1. Enable ImageVolume feature gate in 4.20
+
+    - This feature gate is disabled by default in Kubernetes 1.33, so we need to explicitly enable it for OCP
+
+2. Separate Graduation Approach between OCI Images and OCI Artifacts.
+
+    - **OCI Image Mount**: Promote to GA
+    - **OCI Artifact Mount**: Begin as DevPreview
+        - OCI Artifact functionality is less mature and requires further refinement
+        - Add a new CRI-O option to toggle OCI artifact mount support
+        - Disable OCI artifact mount by default
+
+### Workflow Description
+
+#### OCI Image
+
+Since OCI image mount will be GA in 4.20. No feature gate configuration is needed by users. Creating a pod with volumes.image will pull the image and mount in the container.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+spec:
+  containers:
+  - name: test
+    image: registry.k8s.io/e2e-test-images/echoserver:2.3
+    volumeMounts:
+    - name: volume
+      mountPath: /volume
+  volumes:
+  - name: volume
+    image:
+      reference: quay.io/crio/artifact:v1
+      pullPolicy: IfNotPresent
+```
+
+#### OCI Artifact
+
+While OCI artifact is in dev preview, it fails to mount if a user specifies OCI artifact in reference. To enable the mount, users need to add this drop-in CRI-O configuration.
+
+```
+[crio.image]
+oci_artifact_mount_support = true
+```
+
+When it's enabled, users can mount OCI artifacts in the same way as OCI images. Once we feel comfortable with the CRI-O's implementation and make it GA, we'd remove this option.
+
+### API Extensions
+
+No API change is needed on the OCP side.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+To separate graduation of OCI image and OCI artifact, we are going to add a new option to disable OCI artifact in CRI-O and disable OCI artifact mount by default via MCO.
+
+### Risks and Mitigations
+
+See [KEP-4639: Risks and Mitigations](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4639-oci-volume-source#risks-and-mitigations)
+
+### Drawbacks
+
+## Alternatives (Not Implemented)
+
+See [KEP-4639: Alternatives](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4639-oci-volume-source#alternatives)
+
+- For graduation of OCI artifacts, we could introduce a new field in ContainerRuntimeConfig to toggle it instead of just removing the option.
+
+## Open Questions \[optional\]
+
+## Test Plan
+
+Since OCP just enables the feature gate, the upstream e2e would suffice [https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4639-oci-volume-source\#e2e-tests](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4639-oci-volume-source#e2e-tests). Also it will be tested in CRI-O e2e tests.
```

Comment:
AFAIK we don't run these tests in openshift/kubernetes. I think these tests are node_e2e which are not yet covered in openshift.

Maybe we should add an origin test for this?

Reply:
Since we run upstream e2e on OCP clusters (don't we?), I think e2e tests in upstream is enough. 
https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/image_volume.go

cc @saschagrunert 
https://issues.redhat.com/browse/OCPNODE-3004?focusedId=26936017&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-26936017

Reply:
We don't run tests under test/e2e_node in OCP. I believe we run test/e2e in CI and all the node related tests are skipped (as they require a VM to be created and are not run against an existing K8s cluster).

Those tests require way different setup than what openshift CI expects.

Reply:
Thanks! then we should have e2e tests for OCP as well.